Add missing tests for wallet create, verbose and create_addr

LysanderGG opened the pull request:
**What current issue(s) does this address, or what feature is it adding?**
Add missing tests after PR #720 (also see https://github.com/CityOfZion/neo-python/pull/726#issuecomment-442285227)

The tests are for the following commands `wallet create`, `wallet verbose`, `wallet create_addr`.

**How did you solve this problem?**
Added some tests.

**How did you make sure your solution works?**
Run `make test`.

**Are there any special changes in the code that we should be aware of?**
No.

**Please check the following, if applicable:**

- [x] Did you add any tests?
- [x] Did you run `make lint`?
- [x] Did you run `make test`?
- [x] Are you making a PR to a feature branch or development rather than master?
- [ ] Did you add an entry to `CHANGELOG.rst`? (if not, please do)


Note:

- I couldn't find any other test related to wallet migration so I was unable to test the command `wallet migrate`.

jseagrave21 commented:
This looks good. Could you also add a test for `wallet`?

jseagrave21 commented:
Also, it would be nice to see a test for an invalid `wallet` command

jseagrave21 commented:
And, I was looking here: https://coveralls.io/builds/20405190/source?filename=neo/Prompt/Commands/Wallet.py#L109
Is the logic here wrong? It seems like we have nothing actually triggering `start_wallet_loop`

jseagrave21 commented on neo/Prompt/Commands/tests/test_wallet_commands.py:
```diff
@@ -34,8 +36,73 @@ def GetWallet1(cls, recreate=False):
+                args = ['create', path]
+                self.assertFalse(os.path.isfile(path))
+                res = CommandWallet().execute(args)
+                self.assertEqual(str(type(res)), "<class 'neo.Implementations.Wallets.peewee.UserWallet.UserWallet'>")
```
@LysanderGG based on @ixje's review, you should probably change this test too https://github.com/CityOfZion/neo-python/pull/733/files#diff-f8bf8cf1fb6fe5737ee276e280ad2959R154
I suppose my test set a bad precedent 😳

LysanderGG commented:
@ixje @jseagrave21 Thank you for the first review!

1. test for wallet (coverage)
-> Done, but there's not much to test.
Shall I test what is printed to std::out?

2. test invalid wallet (coverage)
-> Done, but there's not much to test.

3. ensure wallet loop is actually started when creating a new wallet (functionality)
The bug had been introduced with #726.
I fixed it.

Please review a second time when you have time.

ixje commented on neo/Prompt/Commands/tests/test_wallet_commands.py:
```diff
@@ -34,8 +36,84 @@ def GetWallet1(cls, recreate=False):
+            self.assertEqual(res, None)
+
+    def test_wallet_wrong_command(self):
+        with patch('neo.Prompt.PromptData.PromptData.Prompt'):
```
this does not test invalid command. It passes because it has no wallet open and therefore returns `None`.

ixje replied:
if your IDE doesn't support coverage highlighting you could alternatively run `make coverage` or have a look at the `coveralls` output in your PR as a last resort.

ixje commented on neo/Prompt/Commands/tests/test_wallet_commands.py:
```diff
@@ -34,8 +36,84 @@ def GetWallet1(cls, recreate=False):
     # Beginning with refactored tests
 
+    def test_wallet(self):
+        with patch('neo.Prompt.PromptData.PromptData.Prompt'):
```
what are we trying to patch? I don't see it used.

jseagrave21 replied:
Also, one way to make this test less confusing would be `return Prompt data.Wallet` which you could after as true. You could also verify `stdout` if you wanted but not necessary.

jseagrave21 replied:
I also think this test is failing on no wallet

ixje commented on neo/Prompt/Commands/tests/test_wallet_commands.py:
```diff
@@ -93,6 +171,54 @@ def test_wallet_close(self):
+        with patch('neo.Prompt.PromptData.PromptData.Prompt'):
+            # test wallet verbose with no wallet
+            args = ['verbose']
+            res = CommandWallet().execute(args)
```
same as previous. This does not test verbose, but fails on a not having an open wallet.

LysanderGG replied:
This test is specifically testing `test wallet verbose with no wallet`

ixje commented on neo/Prompt/Commands/tests/test_wallet_commands.py:
```diff
@@ -93,6 +171,54 @@ def test_wallet_close(self):
+
+            # test wallet create address with no argument
+            args = ['create_addr']
+            res = CommandWallet().execute(args)
```
not specifically at this line, but in the function it calls (not seeing that code here); 
Can you please give a more user friendly error message and/or capture the "not enough arguments" problem sooner? We now get

```
neo> wallet create_addr
int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```
which is really confusing because you might try interpret this as it indicating that "create_addr" is read as an int() argument

LysanderGG replied:
Amend of ^ 
I removed the early detection and checked the argument in `CommandWalletCreateAddress`.

ixje commented:
> I couldn't find any other test related to wallet migration so I was unable to test the command wallet migrate

Skip it. I'm going to remove all `migrate` code. It's old an no longer used.

jseagrave21 commented on neo/Prompt/Commands/Wallet.py:
```diff
@@ -108,11 +107,10 @@ def execute(self, arguments):
 
```
This case is still not covered. I think at least the anticipated case [here](https://coveralls.io/builds/20433094/source?filename=neo%2FPrompt%2FCommands%2FWallet.py#L101) should be tested.
I am not sure what the expected error is, but @ixje has mentioned to me in previous reviews that catchall try/except methods are undesirable. If this is still desirable logic, then I am okay with leaving it.
Also, if we don't want to cover this case then OK, I am just point it out.
https://coveralls.io/builds/20433094/source?filename=neo%2FPrompt%2FCommands%2FWallet.py#L102

LysanderGG replied:
As a note, this is the copy of the code in the `development` branch.
I can force an exception for the sake of coverage.

jseagrave21 replied:
Right that is not what I'm going for explicitly. What it looks like is this specified case where `os.path.isfile(path)` is `True` isn't tested.

LysanderGG replied:
Got it :+1: , I can make sure to cover these cases.

LysanderGG commented on neo/Prompt/Commands/Wallet.py:
```diff
@@ -108,11 +107,10 @@ def execute(self, arguments):
 
         else:
             print("Please specify a path")
-            return
```
I saw this, but I'm mitigated about whether it should stay as is or not. 
If someone calls `CommandWalletCreate.execute()` directly it would skip the argument validation.

@ixje Any advice?

jseagrave21 replied:
Similar to my previous realization, I understand now that this should be caught by your new logic concerning required input. I think what you have should be good but then we should remove redundant code.

LysanderGG commented on neo/Prompt/Commands/Wallet.py:
```diff
@@ -147,11 +145,9 @@ def execute(self, arguments):
 
         else:
             print("Please specify a path")
-            return
```
Yeah I notified it here: https://github.com/CityOfZion/neo-python/pull/733#discussion_r238490876
I think `construct_send_basic` check should stay as is because it can be called from somewhere else.

jseagrave21 replied:
Ooh. 😊  Sorry, I didn't see that earlier. Okay.

LysanderGG commented:
@ixje @jseagrave21 
I pushed a new commit https://github.com/CityOfZion/neo-python/pull/733/commits/d8d0b04cb2a300b9b9051add1cc63889d3f12ebe
Please have a look.

I changed the tests to use the subcommand directly instead of testing through `CommandWallet`. I think it's better for the unit tests to use the smaller unit possible.

If we want integration tests, we should probably test from the prompt directly instead of `CommandWallet` that is kind of in between.

Please let me know what you think.

Note: testing with the subcommands allows to cover all the code. I added a some edge case detection in the subcommand's `execute()` method where necessary.
In practice, when called through `CommandWallet` it is detected earlier.

jseagrave21 commented:
@LysanderGG I prefer integrated tests which simulate user input as much as possible. When I implemented the testing method here https://github.com/CityOfZion/neo-python/pull/726#issuecomment-441996896 I wasn't sure how to test directly from prompt.py. But if that is possible, I would prefer that.

ixje commented:
Back here https://github.com/CityOfZion/neo-python/pull/726#discussion_r236785324 I kind of reasoned that testing "GroupCommand" vs "SubCommand" doesn't really matter, although "GroupCommand" makes it more obvious what scenario we're working on and will need testing anyway. Which would be in line with @jseagrave21 's last comment.

After your discussion above I kind of realise that people might start calling commands directly when they're extending `neo-python`. e.g. when they implement a web interface, REST API, new RPC commands or want to control neo-python programmatically. They should not have to re-create the logic for let's say `sending assets` when there's a `CommandSend` already doing this. Now being aware of this I don't think we can avoid some code duplication between the "Group Command"  and "SubCommand" so to say. For the extensibility reasons mentioned before, I would now lean towards testing commands directly while not excluding testing group commands. (let me know if that last line doesn't make sense and I'll try to rephrase)

ixje commented:
> I think that the best to do is keep the functionnalities not related to the command line outside of Commands so that these functions can be used elsewhere.

I actually had that in mind (separate presentation from functionality) before writing my previous comment, but thought it might lead to more confusion at this point. 

> I'm a bit lost with this big discussion; can you
No worries, same here. This is a good showcase why we should keep PR's small, because discussions can still get big :)

I think for this one let's get it in and we'll think about moving out the body of `execute()` functions before we do the big merge. That way we can have it in a clean PR which is easier to discuss. @jseagrave21 any objections at this point or things that need solving before merging?

PS: I added the last paragraph as a TODO item to the original "CLI refactoring post"
